src/liboslexec/batched_llvm_gen.cpp

```diff
@@ -6473,6 +6496,217 @@ LLVMGEN (llvm_gen_spline)
+        llvm::Value *mem_void_ptr = rop.ll.GEP (comp_ptr, 0, 2);;
+        mem_void_ptr = rop.ll.ptr_cast(mem_void_ptr, rop.ll.type_void_ptr());
+
+        // For the weighted closures, we need a surrounding "if" so that it's safe
+        // for osl_allocate_weighted_closure_component to return NULL (unless we
+        // know for sure that it's constant weighted and that the weight is
+        // not zero).
+        llvm::BasicBlock *next_block = NULL;
+        llvm::BasicBlock *notnull_block = rop.ll.new_basic_block ("non_null_closure");
+        next_block = rop.ll.new_basic_block ("");
+        llvm::Value *cond = rop.ll.op_ne (rop.ll.ptr_cast(comp_ptr,rop.ll.type_void_ptr()), rop.ll.void_ptr_null());
+        rop.ll.op_branch (cond, notnull_block, next_block);
+        
+        // If the closure has a "prepare" method, call
+        // prepare(renderer, id, memptr).  If there is no prepare method, just
+        // zero out the closure parameter memory.
+        if (clentry->prepare) {
+            // Call clentry->prepare(renderservices *, int id, void *mem)
+            llvm::Value *funct_ptr = rop.ll.constant_ptr((void *)clentry->prepare, rop.llvm_type_prepare_closure_func());
+            llvm::Value *args[] = {render_ptr, id_int, mem_void_ptr};
+            rop.ll.call_function (funct_ptr, args);
+        } else {
+            rop.ll.op_memset (mem_void_ptr, 0, clentry->struct_size, 4 /*align*/);
+        }
+
+        for (int carg = 0; carg < clentry->nformal; ++carg) {
+            const ClosureParam &p = clentry->params[carg];
+
+            if (p.key != NULL) break;
+            OSL_DASSERT(p.offset + p.field_size <= clentry->struct_size);
+            Symbol &sym = *rop.opargsym (op, carg + 2 + weighted);
+
+            bool arg_is_uniform = sym.is_uniform();
+            
+            TypeDesc simpletype (sym.typespec().simpletype());
+            int num_elements = simpletype.numelements();
+            int num_components = simpletype.aggregate;
+
+            llvm::Value *dest_base = rop.ll.offset_ptr (mem_void_ptr, p.offset);
+            dest_base = rop.llvm_ptr_cast(dest_base, p.type);
+
+            for (int a = 0;  a < num_elements;  ++a) {
+                llvm::Value *arrind = simpletype.arraylen ? rop.ll.constant(a) : NULL;
+                for (int c = 0; c < num_components; c++) {
+
+                    llvm::Value *dest_elem;
+                    if (num_components > 1)
+                        dest_elem = rop.ll.GEP(dest_base, a, c);
+                    else 
+                        dest_elem = dest_base;
+
+                    // NOTE:  We don't want any uniform arguments to be
+                    // widened, so our typical op_is_uniform doesn't do what we
+                    // want for this when loading.  So just pass arg_is_uniform
+                    // which will avoid widening any uniform arguments.
+                    llvm::Value* loaded = rop.llvm_load_value (sym, 0, arrind,
```

Comment:
This will reload the value for each lane_index which is probably a performance problem.  Also as noted above alot of code is generated due to loop unrolling.  Printf started being unrolled loop as well, but the amount of IR generated made it difficult to interpret, and we switched to generating a loop, benefits included loading the values only once before the loop and extracting the lane's value within the loop (Delayed Extraction), as similar approach could be made here.

Reply:
Indeed, rewriting this to generate a loop instead of unrolling it and hoisting as much as possible is a good idea for an improvement.  However, I think that is a large enough modification that it should be done in a separate PR.